File: chat-traduction.js
/* chat-traduction.js — TranslationService du chat (feature 22/08).

   Le français s'affiche TOUJOURS immédiatement, la traduction japonaise arrive
   en arrière-plan. Cache par (source|target|texte) : on ne traduit jamais deux
   fois la même phrase. Aucune clé dans le navigateur : le serveur local
   (chat-traduction.py, port 8766) possède DEEPSEEK_API_KEY, ou répond 501.
   Si le serveur est absent/lent/en erreur, un petit dictionnaire de phrases
   courantes sert de secours honnête ; sinon on garde le français, sans casser
   le chat. Le module n'importe RIEN du jeu ni du world builder : le chat et
   le monde 3D restent deux responsabilités séparées.
*/
const URL_TRADUCTION = 'http://127.0.0.1:8766/traduire';
const DELAI_MAX_MS = 4000;

/* ── secours hors-ligne : quelques phrases courantes, japonais naturel ── */
const DICTIONNAIRE_FR_JA = {
  'bonjour': 'こんにちは。',
  'salut': 'やあ。',
  'bonsoir': 'こんばんは。',
  'bonne nuit': 'おやすみ。',
  'au revoir': 'さようなら。',
  'merci': 'ありがとう。',
  'merci beaucoup': 'ありがとうございます。',
  'de rien': 'どういたしまして。',
  'bonne chance': 'がんばってね。',
  'bon appétit': 'いただきます。',
  'comment tu vas': '元気？',
  'comment vas tu': '元気？',
  'ça va': '大丈夫。',
  'ça va je vais rentrer chez moi': '大丈夫。家に帰るよ。',
  'je vais bien': '元気だよ。',
  'je suis perdu': '道に迷った。',
  'je suis fatigue': '疲れた。',
  'j ai faim': 'お腹が空いた。',
  'aide moi': '助けて。',
  'a l aide': '助けて！',
  'attention': '気をつけて。',
  'bonjour vous allez bien': 'こんにちは、お元気ですか？',
  'bonjour bienvenue dans mon monde': 'こんにちは！私の世界へようこそ。',
  'cette maison est magnifique': 'この家は本当に美しいね。',
  'attends moi ici je reviens bientôt': 'ここで待っていて。すぐ戻るから。',
  'attends moi ici je reviens bientot': 'ここで待っていて。すぐ戻るから。',
  'je m appelle': null,   // géré par le modèle ci-dessous
  'je m appelle patrick': '私の名前はパトリックです。',
  'comment tu t appelles': '君の名前は何？',
  'je t aime': '好きだよ。',
  'tu es fort': '君は強いね。',
  'c est beau': 'きれいだね。',
  'c est magnifique': 'すごくきれいだね。',
  'bienvenue': 'ようこそ。',
  'ou est la sortie': '出口はどこ？',
  'par ici': 'こっちだよ。',
  'merci de m aider': '助けてくれてありがとう。',
  'cet endroit est magnifique': 'この場所は本当に美しいね。',
};

function normaliserFR(t) {
  return (t || '').toLowerCase()
    .normalize('NFD').replace(/[\u0300-\u036f]/g, '')   // enlève les accents
    .replace(/[.,!?;:«»"'()…]/g, ' ')                    // ponctuation → espace
    .replace(/\s+/g, ' ').trim();
}

const Traduction = {
  _cache: new Map(),
  _appelsServeur: 0,
  _delaiMaxMs: DELAI_MAX_MS,

  /* translate(text, source, target) — l'interface conceptuelle demandée.
     Ne renvoie JAMAIS d'exception : null = pas de traduction (français gardé). */
  async translate(text, source = 'fr', target = 'ja') {
    const t = (text || '').trim();
    if (!t) return null;                       // 3. message vide
    const cle = source + '|' + target + '|' + t;
    if (this._cache.has(cle)) return this._cache.get(cle);   // 8. cache

    let resultat = null;
    try {
      resultat = await this._servirAvecDelai(t, source, target);
    } catch (e) { resultat = null; }           // 6/7. erreur ou timeout
    if (!resultat) resultat = this._dictionnaire(t) || null;
    if (resultat) this._cache.set(cle, resultat);
    return resultat;
  },

  async _servirAvecDelai(texte, source, target) {
    this._appelsServeur++;
    const promesse = this._serveur(texte, source, target);
    const sablier = new Promise((_, rej) =>
      setTimeout(() => rej(new Error('timeout traduction')), this._delaiMaxMs));
    return Promise.race([promesse, sablier]);
  },

  async _serveur(texte, source, target) {
    const rep = await fetch(URL_TRADUCTION, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text: texte, source, target }),
    });
    if (!rep.ok) return null;
    const d = await rep.json();
    const tr = d && d.traduction;
    return (typeof tr === 'string' && tr.trim()) ? tr.trim() : null;
  },

  _dictionnaire(texte) {
    const n = normaliserFR(texte);
    if (!n) return null;
    if (DICTIONNAIRE_FR_JA[n]) return DICTIONNAIRE_FR_JA[n];
    // une phrase qui CONTIENT un élément connu (ex. « Je m'appelle Lucie. »)
    if (/^je m appelle /.test(n)) {
      const nom = n.replace(/^je m appelle /, '').trim();
      return '私の名前は' + nom + 'です。';
    }
    return null;   // inconnu → le français est conservé (noms propres intacts)
  },

  purgerCache() { this._cache.clear(); },
  get appelsServeur() { return this._appelsServeur; },
};

window.Traduction = Traduction;

File: chat-traduction.py
#!/usr/bin/env python3
# chat-traduction.py — serveur local de traduction du chat (port 8766).
#
# Responsabilité SÉPARÉE du world builder : le chat parle à ce petit serveur,
# qui possède la clé DeepSeek (env DEEPSEEK_API_KEY, jamais dans le code ni
# dans le navigateur). Si la clé est absente, il répond 501 et le jeu garde
# le français (le navigateur a aussi son petit dictionnaire de secours).
# Multi-threadé : un client qui meurt en plein transfert ne fige pas le serveur
# (leçon payée le 22/08 avec serveur-nocache mono-thread).
#
# Lancement :  DEEPSEEK_API_KEY=... python3 chat-traduction.py
# Test :       curl -s -X POST http://127.0.0.1:8766/traduire \
#                -H 'Content-Type: application/json' \
#                -d '{"text":"Bonjour, vous allez bien ?","source":"fr","target":"ja"}'
import http.server
import json
import os
import socketserver
import urllib.error
import urllib.request

PORT = 8766
URL_DEEPSEEK = "https://api.deepseek.com/chat/completions"
MODELE = "deepseek-chat"

PROMPT_SYSTEME = (
    "Tu traduis du français vers un japonais naturel et conversationnel, comme "
    "le parlerait un personnage d'anime. Garde le sens, le ton, les émotions "
    "et les noms propres. N'ajoute RIEN d'autre : réponds uniquement avec la "
    "traduction, sans guillemets ni préfixe."
)


def traduire(texte: str) -> str:
    cle = os.environ.get("DEEPSEEK_API_KEY", "").strip()
    if not cle:
        raise OSError("clé DEEPSEEK_API_KEY absente")
    corps = json.dumps({
        "model": MODELE,
        "messages": [
            {"role": "system", "content": PROMPT_SYSTEME},
            {"role": "user", "content": "Traduis en japonais : " + texte},
        ],
        "temperature": 0.3,
    }).encode("utf-8")
    req = urllib.request.Request(
        URL_DEEPSEEK, data=corps,
        headers={"Content-Type": "application/json", "Authorization": "Bearer " + cle},
        method="POST")
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            reponse = json.loads(r.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        raise OSError("DeepSeek HTTP %s : %s" % (e.code, e.read()[:200]))
    except Exception as e:
        raise OSError("DeepSeek injoignable : %s" % e)
    try:
        return reponse["choices"][0]["message"]["content"].strip()
    except (KeyError, IndexError, TypeError):
        raise OSError("réponse DeepSeek inattendue")


class H(http.server.BaseHTTPRequestHandler):
    def _json(self, code, obj):
        data = json.dumps(obj, ensure_ascii=False).encode("utf-8")
        self.send_response(code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def do_POST(self):
        if self.path != "/traduire":
            self._json(404, {"erreur": "route inconnue : " + self.path})
            return
        try:
            n = int(self.headers.get("Content-Length") or 0)
            corps = json.loads(self.rfile.read(n) or b"{}")
        except (ValueError, json.JSONDecodeError):
            self._json(400, {"erreur": "JSON invalide"})
            return
        texte = (corps.get("text") or "").strip()
        if not texte:
            self._json(400, {"erreur": "texte vide"})
            return
        try:
            traduction = traduire(texte)
        except OSError as e:
            self._json(501, {"erreur": str(e)})
            return
        self._json(200, {"traduction": traduction, "source": "deepseek"})

    def log_message(self, *a):
        pass


if __name__ == "__main__":
    socketserver.TCPServer.allow_reuse_address = True
    print("Serveur de traduction sur http://127.0.0.1:%d" % PORT)
    socketserver.ThreadingTCPServer(("", PORT), H).serve_forever()
